protocols/rendezvous/src/server.rs

```diff
@@ -109,190 +113,169 @@ pub enum Event {
+
+                Some(InboundOutEvent::NewRegistration(new_registration))
             }
-        }
-        inbound::OutEvent::DiscoverRequested {
-            namespace,
-            cookie,
-            limit,
-        } => match registrations.get(namespace, cookie, limit) {
-            Ok((registrations, cookie)) => {
-                let discovered = registrations.cloned().collect::<Vec<_>>();
-
-                vec![
-                    NetworkBehaviourAction::NotifyHandler {
-                        peer_id,
-                        handler: NotifyHandler::One(connection),
-                        event: handler::InboundInEvent::NotifyInboundSubstream {
-                            id,
-                            message: inbound::InEvent::DiscoverResponse {
-                                discovered: discovered.clone(),
-                                cookie,
-                            },
-                        },
-                    },
-                    NetworkBehaviourAction::GenerateEvent(Event::DiscoverServed {
-                        enquirer: peer_id,
-                        registrations: discovered,
-                    }),
-                ]
+            Message::Unregister(namespace) => {
+                substream.close().await?;
+
+                Some(InboundOutEvent::Unregister(namespace))
             }
-            Err(_) => {
-                let error = ErrorCode::InvalidCookie;
-
-                vec![
-                    NetworkBehaviourAction::NotifyHandler {
-                        peer_id,
-                        handler: NotifyHandler::One(connection),
-                        event: handler::InboundInEvent::NotifyInboundSubstream {
-                            id,
-                            message: inbound::InEvent::DeclineDiscoverRequest(error),
-                        },
-                    },
-                    NetworkBehaviourAction::GenerateEvent(Event::DiscoverNotServed {
-                        enquirer: peer_id,
-                        error,
-                    }),
-                ]
+            Message::Discover {
+                namespace,
+                cookie,
+                limit,
+            } => match registrations.get(namespace, cookie, limit) {
+                Ok((registrations, cookie)) => {
+                    substream
+                        .send(Message::DiscoverResponse(Ok((
+                            registrations.cloned().collect(),
+                            cookie,
+                        ))))
+                        .await?;
+                    substream.close().await?;
+
+                    None
+                }
+                Err(e) => {
+                    substream
+                        .send(Message::DiscoverResponse(Err(ErrorCode::InvalidCookie)))
+                        .await?;
+                    substream.close().await?;
+
+                    None
+                }
+            },
```

Comment:
This is where we harvest the fruits of this work! Because the `ConnectionHandler` has access to `RegistrationData`, we can "answer" the remote's query inline which allows us to use `async/await` to write the protocol. In the end, it is just a simple request-response but we couldn't do this before because we needed to suspend the stream and wait for the data coming back from the behaviour!

I think this could be a good fit for any protocol where all the data nicely fits into memory.

Reply:
> I think this could be a good fit for any protocol where all the data nicely fits into memory.

Note that with the latest patches, the data is shared between all connections with an `Arc` and thus not actually duplicated in-memory.